Vim mode

idealemu opened the pull request:
<!--
Please check the following before you make a pull request:

* The branch you are making the pull request from (in your own fork) has a unique and descriptive
  name. Do not make a pull request directly from your copy of `main`.
* If you are submitting translation files, only submit the files that you have added translations
  to, not the other .ts files that may have been updated in the process.
* Make sure your contribution follows the style guide and other requirements mentioned in the
  CONTRIBUTING.md file in the repository.
* Fill in the Summary section below, and if relevant, mention the issue numbers related to the PR.
-->

**Summary:**
Adding an editor Vim-like mode to novelWriter. This will enable users who activate this feature to navigate files using basic vim motions, performing basic editing in a vim "Normal mode" and write text normally in the "Insert mode".

**Related Issue(s):**
https://github.com/vkbo/novelWriter/issues/2493
**Reviewer's Checklist:**

* [ ] The header of all files contain a reference to the repository license
* [ ] The overall test coverage is increased or remains the same as before
* [ ] All tests are passing
* [ ] All linting checks are passing and the style guide is followed
* [ ] Documentation (as docstrings) is complete and understandable
* [ ] Only files that have been actively changed are committed

vkbo commented:
Just an FYI (I already fixed it) under "Related Issue(s)" you put "Closes" and the issue number (not the link). It then triggers the automatic linking to the ticket.

vkbo commented on novelwriter/gui/doceditor.py:
```diff
@@ -919,6 +934,10 @@ def keyPressEvent(self, event: QKeyEvent) -> None:
         self._lastActive = time()
+
+        if self._handleVimKeyPress(event):
+            return
```
Could we check `CONFIG.vimModeEnabled` here instead of inside the function so that we skip the function call if it isn't enabled? There is some overhead for calling functions and it's best to avoid it if it isn't necessary.

Something like:
```python
if CONFIG.vimModeEnabled and self._handleVimKeyPress(event):
    return
```

vkbo replied:
Hmm, it seems the function always returns True if Vim Mode is enabled, so maybe the whole boolean return can be dropped and instead use:

```python
if CONFIG.vimModeEnabled:
    self._handleVimKeyPress(event)
    return
```

vkbo commented:
This is looking really great. I'll see if I can fix my part of it this weekend.

Test coverage is still a little spotty. You can click forward to the test coverage report from the codecov check here on the pull request, or if you run tests locally with `run_tests.py`, setting the `-r` flag will generate a HTML report folder at the root of the repo. On Linux, the coverage report should show 26 missing lines when the code is covered 100%. The remaining lines are covered on Windows and MacOS.

vkbo commented:
I'm also happy to write some of the tests. It helps to understand the code flow. I also find that writing tests is a good way to test the logic of the code. I often discover issues and corner cases when writing tests.

idealemu commented:
> This is looking really great. I'll see if I can fix my part of it this weekend.
> 
> Test coverage is still a little spotty. You can click forward to the test coverage report from the codecov check here on the pull request, or if you run tests locally with `run_tests.py`, setting the `-r` flag will generate a HTML report folder at the root of the repo. On Linux, the coverage report should show 26 missing lines when the code is covered 100%. The remaining lines are covered on Windows and MacOS.

Super ! I did a trial usage of it yesterday evening and I have one or two features I realised I was missing. I'll bump the test coverage up once those are added. :smile:

vkbo commented:
Not sure if you've noticed, but the codecov job links through to a report that shows what lines are not covered by tests. Only 7 lines are missed now!

https://app.codecov.io/gh/vkbo/novelWriter/pull/2499

idealemu commented:
> I fixed the isort linting complaint. It's a little clunky to set up because ruff and isort interfere a little with each other. Unfortunately, import sorting in ruff is a litt under-configurable.

Yes, I couldn't quite figure out what was wrong with it. Thanks !

> In any case, the PR looks complete to me in terms of features. I can do the parts I need to do in a follow up PR. It may be cleaner.

In my eyes almost ! I have two or three other basic commands I would like to add before calling it complete (and corresponding test coverage), and I would like to refactor everything a little, but I could do the refactor in another PR if you want ? It would just be refactor of the _handleVim method, and of the test cases which currently are really just "add everything until coverage is hit" without much organisation.

vkbo commented:
> Yes, I couldn't quite figure out what was wrong with it. Thanks !

If you use VSCode, and have Microsoft's Isort extension installed, add this config to the `settings.json` file for the workspace or `.vscode` folder:

```json
    "[python]": {
        "editor.defaultFormatter": "ms-python.isort",
        "editor.codeActionsOnSave": {
            "source.organizeImports": "always",
            "source.fixAll": "always"
        },
    },
```

The needed settings are in `pyproject.toml`, so the additional config above should ensure that the imports are auto-fixed when you save.

You can also run:

```bash
isort --check .
```

I'm considering enabling Ruff formatting for the code base, but it lacks a few features I need, so I haven't done so yet.

> In my eyes almost ! I have two or three other basic commands I would like to add before calling it complete (and corresponding test coverage), and I would like to refactor everything a little, but I could do the refactor in another PR if you want ? It would just be refactor of the _handleVim method, and of the test cases which currently are really just "add everything until coverage is hit" without much organisation.

Sure, no rush.

vkbo commented on novelwriter/gui/doceditor.py:
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
+        self._normalCommand = ""
+        self._visualCommand = ""
+
+    def setMode(self, new_mode: nwVimMode) -> None:
+        """Switch vim mode."""
+        self._mode = new_mode
+        self.resetCommand()
```
```suggestion
    def setMode(self, newMode: nwVimMode) -> None:
        """Switch vim mode."""
        self._mode = newMode
        self.resetCommand()
```

vkbo commented on novelwriter/gui/doceditor.py:
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
+    )
+
+    def __init__(self) -> None:
+        self.PREFIX_KEYS = ["d", "y", "g", "z"]
+        self.VISUAL_PREFIX_KEYS = ["g"]
+        self.SUFFIX_KEYS = ["w"]
```
As far as I can tell, these lists can just be changed to strings. A string is also classified as a sequence in Python. As long as you only check a single character, there's no need to make them lists.

```suggestion
        self.PREFIX_KEYS = "dygz"
        self.VISUAL_PREFIX_KEYS = "g"
        self.SUFFIX_KEYS = "w"
```

vkbo commented on novelwriter/gui/doceditor.py:
```diff
@@ -3220,3 +3585,74 @@ def updateMainCount(self, count: int, selection: bool) -> None:
+        self._mode = new_mode
+        self.resetCommand()
+
+    def getMode(self) -> nwVimMode:
+        """Return current vim mode."""
+        return self._mode
```
Generally, I try to avoid get methods if they are plain and simple like this one and instead use Python property instead.

```suggestion
    @property
    def mode(self) -> nwVimMode:
        """Return current vim mode."""
        return self._mode
```

vkbo commented on novelwriter/gui/doceditor.py:
```diff
@@ -948,6 +968,337 @@ def keyPressEvent(self, event: QKeyEvent) -> None:
+            self.setVimMode(nwVimMode.VISUAL)
+            cursor = self.textCursor()
+            self._visualAnchor = cursor.position()
+            cursor.movePosition(cursor.MoveOperation.Right, cursor.MoveMode.KeepAnchor, 1)
```
These enums should be extracted from `QTextCursor` itself, not the instance of the cursor. It makes them easier to find later, and it's consistent with the rest of the code.

Also, most of the enums that are frequently accessed are defined with a short name variable in `types.py`. Many of the ones you use are already defined there. Feel free to add more if you find they are used often.

I added these shortnames when I ported novelWriter from PyQt5 to PyQt6. Since PyQt6 you have to use the full namespace path of the enum values, which means they tend to be very long, and since these values have to be looked up on each dot, is a bit of a waste of cycles.

I'm trying to be very conscious of performance on all code that is triggered on key press as Python is relatively slow, and bouncing calls between Python and the Qt library has a lot of overhead due to type conversion back and forth. There are already indications that some users are experiencing typing lag which may or may not be due to this. There are also typing lag issues related to OS rendering of the text, which is unrelated, but I'm trying to not add to the problem.

vkbo commented:
Hi again. I've updated the PR and fixed the bits I have mentioned and a bunch of other things for consistency with the rest of the code. It looks fine to me now. The `vimmode` variable is also now present in the `novelwriter.conf` file, but not added in Preferences yet. I'll do that in a separate PR later.,

However, a major problem for me is that I don't understand how Vim Mode actually works, so it's difficult for me to test that my changes didn't break anything. Could you give it a test before we merge it?

vkbo commented:
Great, thanks!

I have a performance optimisation branch as well where I only read the `self._vim.command` value once and cache it for the rest of the if-elif. As far as I can tell from the code, that shouldn't be a problem. The only exception from the pattern is the one elif that doesn't return `True`. If that's just something you forgot, my rewrite should be fine. (It passes your tests.)

idealemu commented on novelwriter/gui/doceditor.py:
```diff
@@ -1879,6 +1909,328 @@ def _insertCommentStructure(self, style: nwComment) -> None:
+            cursor.movePosition(QtMoveUp)
+            cursor.endEditBlock()
+            self.setTextCursor(cursor)
+            self.setVimMode(nwVimMode.INSERT)
```
I'm 99% sure this is just that I forgot it.


Currently because of the behaviour of `_handleVimKeyPress` if it would not have the return True it would just move to the next statement which is nothing in this case. 

```python
    def _handleVimKeyPress(self, event: QKeyEvent) -> None:
        """Handle key events for Vim mode.
        If vim mode is not enabled, typing behaves as normal.
        """
        # --- INSERT mode, bypass ---
        if self._vim.mode == nwVimMode.INSERT:
            super().keyPressEvent(event)
            return  # Normal typing

        if self._handleVimNormalModeModeSwitching(event):
            return

        if self._handleVimVisualMode(event):
            return

        if self._handleVimNormalMode(event):
            return

        return
```

Could be simplified to:
```python
    def _handleVimKeyPress(self, event: QKeyEvent) -> None:
        """Handle key events for Vim mode.
        If vim mode is not enabled, typing behaves as normal.
        """
        # --- INSERT mode, bypass ---
        if self._vim.mode == nwVimMode.INSERT:
            super().keyPressEvent(event)
            return  # Normal typing

        if self._handleVimNormalModeModeSwitching(event):
            return

        if self._handleVimVisualMode(event):
            return

        self._handleVimNormalMode(event)  # Don't care about return since it's the last call
```

But we would still want early return statements in the handleVim methods to stop evaluating all the if statements :)

idealemu commented:
I have added `docs/vim-mode.md` with basic documentation.

I tested while doing some writing and everything seems to work fine.
However, I also noticed that there are two commands "db" and "de" are missing (delete back, and delete end, which are both analogous to delete word).

Based on existing code I would suggest:

```python

if self._vim.command == "db":
    cursor.beginEditBlock()
    cursor.movePosition(QtMovePreviousWord, QtKeepAnchor)
    self._vim.yankToInternal(cursor.selectedText())
    cursor.removeSelectedText()
    cursor.endEditBlock()
    self.setTextCursor(cursor)
    self._vim.resetCommand()
    return True

if self._vim.command == "de":
    cursor.beginEditBlock()
    # Extend selection to end of current/next word
    origPos = cursor.position()
    cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
    if cursor.position() == origPos:               # already at end-of-word
        textLen = len(self.toPlainText())
        if origPos < textLen:
            cursor.movePosition(QtMoveNextChar, QtKeepAnchor)
            cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
    self._vim.yankToInternal(cursor.selectedText())
    cursor.removeSelectedText()
    cursor.endEditBlock()
    self.setTextCursor(cursor)
    self._vim.resetCommand()
    return True
```
But I need to seem if this matches the performances changes you made :smiley: 
Otherwise I think I am decently satisfied with what this PR accomplishes.

vkbo commented:
Don't worry about the performance changes. I can adapt those to fit.

I see there are merge conflict issues with the current main branch. I can fix those if you wish, once you're done with your changes.

idealemu commented:
Super ! I just added these two, and corresponding test coverage and doc update then :) If you could take care of the conflicts that would be great ! Docs are currently markdown as you suggested, I wanted to learn the format you use but I frankly kind of just want this PR to be ready :smile:

idealemu commented:
> Updated your branch and fixed linting errors.
> 
> It looks good. One minor thing is that the "Already at end-of-word" condition here is not covered by tests.
> 
> ```python
>         if self._vim.command == "de":
>             cursor.beginEditBlock()
>             # Extend selection to end of current/next word
>             origPos = cursor.position()
>             cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
>             if cursor.position() == origPos:  # Already at end-of-word
>                 textLen = len(self.toPlainText())
>                 if origPos < textLen:
>                     cursor.movePosition(QtMoveNextChar, QtKeepAnchor)
>                     cursor.movePosition(QtMoveEndOfWord, QtKeepAnchor)
>             self._vim.yankToInternal(cursor.selectedText())
>             cursor.removeSelectedText()
>             cursor.endEditBlock()
>             self.setTextCursor(cursor)
>             self._vim.resetCommand()
>             return True
> ```

I just added the test coverage for that :+1:  And expanded the doc on a slight difference in my implementation of the word end command "e" compared to how vim handles it.

vkbo commented:
> For me this is now at a state where I am moving my personal notes and world-building from markdown documents I used to edit in vim into novelWriter to have everything in one place. It might need more features (read more vim motions/commands) in the future, but it good enough for me to dog-food it for my personal novel writing 😄

😄 

> Unless you have other points that I should address I'm willing to consider this PR ready.

Nope. All good! I'm planning to do an RC1 release not this weekend, but next. So I'll merge it before this so it's included in RC1. After that it's usually two weeks of testing before I do the full 2.8 release.

> (Now if novelWriter had the ability to have two editor windows open in a split I would be one happy novel writer !)

Yeah, I'm missing that too sometimes when I'm writing but need to update something in my notes on the right side. I've thought about merging the editor and viewer classes into one and instead make it a edit/view mode in one panel. That way, you can switch either of them between edit and view as needed.

It's non-trivial but it may be worth the effort.
